Archive/Imo/Imo1985Q2.lean

```diff
@@ -0,0 +1,100 @@
+/-
+Copyright (c) 2025 Jeremy Tan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jeremy Tan, David Renshaw
+-/
+import Mathlib.Algebra.Order.Ring.Canonical
+import Mathlib.Algebra.Order.Ring.Star
+import Mathlib.Data.Nat.ModEq
+
+/-!
+# IMO 1985 Q2
+
+Each of the numbers in the set $N=\{1, 2, 3, \dots, n-1\}$, where $n ≥ 3$, is colored with one of
+two colors, say red or black, so that:
+
+1. $i$ and $n-i$ always receive the same color, and
+2. for some $j ∈ N$ relatively prime to $n$, $i$ and $|j-i|$ receive the same color
+for all $i ∈ N, i ≠ j$.
+
+Prove that all numbers in $N$ must receive the same color.
+
+# Solution
+
+Let $a \sim b$ denote that $a$ and $b$ have the same color.
+Because $j$ is coprime to $n$, every number in $N$ is of the form $kj\bmod n$ for a unique
+$1 ≤ k < n$, so it suffices to show that $kj\bmod n \sim (k-1)j\bmod n$ for $1 < k < n$.
+In this range of $k$, $kj\bmod n ≠ j$, so
+
+* if $kj\bmod n > j$, $kj\bmod n \sim kj\bmod n - j = (k-1)j\bmod n$ using rule 2;
+* if $kj\bmod n < j$, $kj\bmod n \sim j - kj\bmod n \sim n - j + kj\bmod n = (k-1)j\bmod n$
+using rule 2 then rule 1.
+-/
+
+namespace Imo1985Q2
+
+open Nat
+
+/-- The conditions on the problem's coloring `C`.
+Although its domain is all of `ℕ`, we only care about its values in `Set.Ico 1 n`. -/
+def Condition (n j : ℕ) (C : ℕ → Fin 2) : Prop :=
+  (∀ i ∈ Set.Ico 1 n, C i = C (n - i)) ∧ (∀ i ∈ Set.Ico 1 n, i ≠ j → C i = C (j - i : ℤ).natAbs)
+
+lemma Int.natAbs_sub_nat_of_lt {a b : ℕ} (h : b ≤ a) : (a - b : ℤ).natAbs = a - b := by
+  omega
+
+lemma Int.natAbs_sub_nat_of_gt {a b : ℕ} (h : a ≤ b) : (a - b : ℤ).natAbs = b - a := by
+  omega
+
+lemma Nat.mod_sub_comm {a b n : ℕ} (h : b ≤ a % n) : a % n - b = (a - b) % n := by
+  rcases n.eq_zero_or_pos with rfl | hn; · simp only [mod_zero]
+  nth_rw 2 [← div_add_mod a n]; rw [Nat.add_sub_assoc h, mul_add_mod]
+  exact (mod_eq_of_lt <| (sub_le ..).trans_lt (mod_lt a hn)).symm
```

Comment:
I'd say the `Int` and `Nat` lemmas belong in files about `Int` and `Nat` rather than in this archive file (and thus belong in namespaces `_root_.Int` and `_root_.Nat` rather than `Imo1985Q2.Int` and `Imo1985Q2.Nat`).

Reply:
I moved the `Int` lemmas to the place you suggested and the `Nat` lemma to `Data.Nat.ModEq`.